Enter your name below in formate of "First_name Last_name"

In [11]:

import os


NameError: name 'ls' is not defined

---

Project - 1

- To download data, go to this link and click download from data tab. You need to have Kaggle account for this.
https://www.kaggle.com/jboysen/us-perm-visas

Instructions for deliverables:
- Do not change anything in any other cell except in cells with comment "# YOUR CODE HERE" or "YOUR ANSWER HERE". If you need to play around with data, play around with it in another notebook. But donot submit that extra work as part of this notebook.
- Do not rename this notebook. Submit file as project1.ipynb only.
- Submit only one notebook. Before submitting, make sure everything runs as expected. To check that, restart the kernel (in the menubar, select Kernel > Restart) and then run all cells (in the menubar, select Cell > Run All).

Instructions for coding:
- Put the dataset in same folder as this file.
- Do not change the name of the dataset. Keep it as "us_perm_visas.csv" only.
- Use groupby in almost all questions.
- Use column 'case_number' for counting purposes. This means if this column has missing value, we don't count that row even if other columns have values in them.
- If same data is avaialble in multiple columns, use column with less missing values.
- Ignore missing data.
- Your output should show exactly what has been asked. Do not show whole big table with all the information. e.g. if it has been asked to list top 3 countries, your out put should have only 3 countries and not list of all countries.

Hint : All questions can be answered by using only one line of code. So try to use one line of code only to get answer.

In [1]:
import warnings
warnings.simplefilter('ignore')
import pandas as pd
data = pd.read_csv("us_perm_visas.csv")

In [2]:
data.head()

add_these_pw_job_title_9089 agent_city agent_firm_name agent_state  \
0                         NaN        NaN             NaN         NaN   
1                         NaN        NaN             NaN         NaN   
2                         NaN        NaN             NaN         NaN   
3                         NaN        NaN             NaN         NaN   
4                         NaN        NaN             NaN         NaN   

  application_type        case_no case_number case_received_date case_status  \
0             PERM  A-07323-97014         NaN                NaN   Certified   
1             PERM  A-07332-99439         NaN                NaN      Denied   
2             PERM  A-07333-99643         NaN                NaN   Certified   
3             PERM  A-07339-01930         NaN                NaN   Certified   
4             PERM  A-07345-03565         NaN                NaN   Certified   

  class_of_admission              ...               ri_pvt_employment_firm_to  \
0                J-1              ...                                     NaN   
1                B-2              ...                                     NaN   
2               H-1B              ...                                     NaN   
3                B-2              ...                                     NaN   
4                L-1              ...                                     NaN   

  ri_us_workers_considered schd_a_sheepherder     us_economic_sector  \
0                      NaN                NaN                     IT   
1                      NaN                NaN  Other Economic Sector   
2                      NaN                NaN              Aerospace   
3                      NaN                NaN  Other Economic Sector   
4                      NaN                NaN           Advanced Mfg   

  wage_offer_from_9089 wage_offer_to_9089 wage_offer_unit_of_pay_9089  \
0                75629                NaN                          yr   
1                37024                NaN                          yr   
2                47923                NaN                          yr   
3                10.97                NaN                          hr   
4               100000                NaN                          yr   

  wage_offered_from_9089 wage_offered_to_9089  wage_offered_unit_of_pay_9089  
0                    NaN                  NaN                            NaN  
1                    NaN                  NaN                            NaN  
2                    NaN                  NaN                            NaN  
3                    NaN                  NaN                            NaN  
4                    NaN                  NaN                            NaN  

[5 rows x 154 columns]

In [3]:
data.shape

(374362, 154)

In [4]:
for i in data.columns:
    print(i)

add_these_pw_job_title_9089
agent_city
agent_firm_name
agent_state
application_type
case_no
case_number
case_received_date
case_status
class_of_admission
country_of_citizenship
country_of_citzenship
decision_date
employer_address_1
employer_address_2
employer_city
employer_country
employer_decl_info_title
employer_name
employer_num_employees
employer_phone
employer_phone_ext
employer_postal_code
employer_state
employer_yr_estab
foreign_worker_info_alt_edu_experience
foreign_worker_info_birth_country
foreign_worker_info_city
foreign_worker_info_education
foreign_worker_info_education_other
foreign_worker_info_inst
foreign_worker_info_major
foreign_worker_info_postal_code
foreign_worker_info_rel_occup_exp
foreign_worker_info_req_experience
foreign_worker_info_state
foreign_worker_info_training_comp
foreign_worker_ownership_interest
foreign_worker_yr_rel_edu_completed
fw_info_alt_edu_experience
fw_info_birth_country
fw_info_education_other
fw_info_postal_code
fw_info_rel_occup_exp
fw_info

Q1. What is the count of applications submitted by each country?

In [5]:
# YOUR CODE HERE
data.groupby('country_of_citizenship').case_number.count()
#use code mentioned below if unique values are required
#data.groupby('country_of_citizenship').apply(lambda x: len(x['case_number'].unique()))

country_of_citizenship
AFGHANISTAN                   16
ALBANIA                       93
ALGERIA                       29
ANDORRA                        1
ANGOLA                        13
ANTIGUA AND BARBUDA            9
ARGENTINA                    760
ARMENIA                      107
ARUBA                          0
AUSTRALIA                   1092
AUSTRIA                      143
AZERBAIJAN                    46
BAHAMAS                       68
BAHRAIN                        7
BANGLADESH                   700
BARBADOS                      29
BELARUS                      329
BELGIUM                      150
BELIZE                        25
BENIN                         18
BERMUDA                        6
BHUTAN                         6
BOLIVIA                      113
BOSNIA AND HERZEGOVINA        36
BOTSWANA                      13
BRAZIL                      1900
BRITISH VIRGIN ISLANDS         2
BRUNEI                         3
BULGARIA                     377
BURKINA FASO        

Q2. List top 3 countries with highest number of applications and also show number of applications by them.

In [7]:
# YOUR CODE HERE
data.groupby('country_of_citizenship').case_number.count().nlargest(3)

country_of_citizenship
INDIA          130707
CHINA           20177
SOUTH KOREA     18209
Name: case_number, dtype: int64

Q3. How many applications were denied?

In [8]:
# YOUR CODE HERE
data[data.case_status=='Denied'].groupby('case_status').case_number.count()

case_status
Denied    12781
Name: case_number, dtype: int64

Q4. How many applicants holding H-1B visas were certified for Permenant Residency? (Consider all forms used for writing "H1-B" in data. Consider only H1-B, not H1-B1)

In [9]:
# YOUR CODE HERE
#I have not considered 'H1B' and 'AOS/H1B' as it is mentioned to use only 'H1-B'
data[(data['class_of_admission'].isin (['H-1B'])) & (data['case_status'] == 'Certified')]['case_status'].value_counts()
#data[(data['class_of_admission'].isin (['H-1B','H1B','AOS/H1B'])) & (data['case_status'] == 'Certified')]['case_status'].value_counts()

Certified    139312
Name: case_status, dtype: int64

Q5. Sort columns in descending order according to number of missing values in them.

In [10]:
# YOUR CODE HERE
data.isnull().sum().sort_values(ascending = False)

orig_file_date                         374081
orig_case_no                           374068
recr_info_job_fair_to                  372569
recr_info_job_fair_from                372564
recr_info_on_campus_recr_to            372497
recr_info_on_campus_recr_from          372490
ji_live_in_dom_svc_contract            371964
foreign_worker_info_education_other    370422
job_info_training_field                369883
job_info_training_num_months           369583
ri_coll_teach_select_date              368632
ri_coll_tch_basic_process              367349
recr_info_coll_teach_comp_proc         367341
job_info_education_other               366593
job_info_alt_combo_ed_other            365148
recr_info_pro_org_advert_to            360292
recr_info_pro_org_advert_from          360286
pw_source_name_other_9089              358603
ri_pvt_employment_firm_to              357922
ri_pvt_employment_firm_from            357920
ri_us_workers_considered               357129
naics_2007_us_title               

Q6. List top 10 cities from which employers have submitted the most applications. How many applicarions were submitted by them?

In [11]:
# YOUR CODE HERE
data.groupby('employer_city').case_number.count().nlargest(10)

employer_city
COLLEGE STATION    11632
NEW YORK            9844
SANTA CLARA         6966
SAN JOSE            6073
MOUNTAIN VIEW       6028
REDMOND             4901
SAN FRANCISCO       4125
SUNNYVALE           4075
PLANO               3844
CHICAGO             3784
Name: case_number, dtype: int64

Q7. Which are the top 10 employers who have submitted the highest number of applications? How many?

In [12]:
# YOUR CODE HERE
data.groupby('employer_name').case_number.count().nlargest(10)

employer_name
COGNIZANT TECHNOLOGY SOLUTIONS US CORPORATION    11587
MICROSOFT CORPORATION                             4625
INTEL CORPORATION                                 4127
GOOGLE INC.                                       3771
AMAZON CORPORATE LLC                              2716
CISCO SYSTEMS, INC.                               2097
INFOSYS LTD.                                      1795
APPLE INC.                                        1730
ORACLE AMERICA, INC.                              1552
FACEBOOK, INC.                                    1268
Name: case_number, dtype: int64

Q8. What is the average time of processing the applications?

In [13]:
# YOUR CODE HERE
(pd.to_datetime(data['decision_date']) - pd.to_datetime(data['case_received_date'])).fillna(0).agg(['mean'])
#if we dont want to use fillna then use the code mentioned below
#(pd.to_datetime(data['decision_date']) - pd.to_datetime(data['case_received_date'])).agg(['mean'])


mean   132 days 21:28:21.691945
dtype: timedelta64[ns]

Q9. What is count of applications for each case status(e.g. Certified, Denied etc) for applicants with "Doctorate" degree?

In [14]:
# YOUR CODE HERE
pd.crosstab(data.foreign_worker_info_education=='Doctorate', data.case_status).iloc[1,:]

case_status
Certified            5965
Certified-Expired    5449
Denied                345
Withdrawn             484
Name: True, dtype: int64

Q.10 What are the top 3 economic sectors that has maximum applicants? How many applications are submitted from these sectors? (Use case_no column here, instead of case_number)

In [15]:
# YOUR CODE HERE
data.groupby('us_economic_sector').case_no.count().nlargest(3)
#use code mentioned below if unique values are required
#data.groupby('us_economic_sector').apply(lambda x: len(x['case_no'].unique())).nlargest(3) 

us_economic_sector
IT                       52617
Advanced Mfg             18749
Other Economic Sector    15798
Name: case_no, dtype: int64

Q.11 What is the mean wage of all the certified visa applications?

In [6]:
# YOUR CODE HERE
#As it was mentioned to ignore missing data I have not imputed missing values
data.groupby('case_status').wage_offered_from_9089.mean()

case_status
Certified             98359.286042
Certified-Expired     96635.161035
Denied                77325.285986
Withdrawn            100241.062797
Name: wage_offered_from_9089, dtype: float64

Q.12 Show one instresting finding of your own which is not covered above. Show the code and write your finding as comment.

In [17]:
# YOUR CODE HERE
#Part A :- Top 5 job titles who got Certified.
part_a = data[data.case_status=='Certified'].groupby('pw_soc_title').case_number.count().nlargest(5)
# Part B :- Top 5 foreign worker major field of study who got certfied
part_b = data[data.case_status=='Certified'].groupby('foreign_worker_info_major').case_number.count().nlargest(5)

print(part_a)
print(part_b)

pw_soc_title
Software Developers, Applications            41201
Computer Systems Analysts                    11065
Software Developers, Systems Software         8467
Electronics Engineers, Except Computer        3910
Computer and Information Systems Managers     2767
Name: case_number, dtype: int64
foreign_worker_info_major
COMPUTER SCIENCE           13895
ELECTRICAL ENGINEERING      6699
COMPUTER ENGINEERING        6398
BUSINESS ADMINISTRATION     5775
MECHANICAL ENGINEERING      3483
Name: case_number, dtype: int64


Q.13 Write a report on your analysis of this dataset. (Around 200 words, but quality of your report will be taken into account more than count of words)

#Report
* Source used :- https://www.kaggle.com/jboysen/us-perm-visas/home

The dataset is about US Permanent Visa Application from 2012-2017 where permanent labor certification is issued to foreign workers to work permanently in United States. The dataset consists of around 154 columns which contains information regarding to foreign worker such as employer, position, wage offered, employee education etc. Analysis of the dataset was done using Pandas library in python. We have ignored missing values and considered the column with least number of missing values when there are duplicate columns in dataset. Finding from Data Analysis performed above: -
1)	Maximum number of applications were submitted for workers having citizen of India, China and South Korea.
2)	Of the total applications 12781 application were denied with top 3 class of admission as H1-B, F-1, and Not in USA population.
3)	Of the total applications submitted by H1-B holders only 139312 got certified for Permanent residency. 
4)	Most applications were submitted from employers at California, Texas, New York respectively.
5)	Cognizant Technology, Microsoft, Intel, Google, Amazon were the top employers to file for the application.
6)	Maximum number of permanent residencies were given to workers with education as Doctorate followed by Master’s and Bachelor’s respectively.
7)	IT is the top economic sector in USA from where maximum number of applications were submitted.
8)	Maximum number of residencies were offered to workers with job title as Software Developers, Electronic Engineers, and Information System Mangers.
9)	The main area of study for foreign workers were Computer Science, Electrical Engineering, Computer Engineering, Business Administration, and Mechanical Engineering.
10)	From information mentioned in points 8 and 9 it can be inferred that workers with STEM degrees were given more preference while certifying for permanent residency in USA.

